Add support for MFA with Duo's Universal Prompt

0x0fbc opened the pull request:
### Overview

This adds support for MFA using Duo's "Universal Prompt".

On March 30th, 2024, the Duo MFA integration that Vaultwarden uses, the 'Traditional Prompt' was end-of-lifed. While Duo's API continues to allow the traditional prompt to be used in some cases, it is entirely unsupported, and it will eventually stop working. Some Vaultwarden users have already begun reporting that their Duo API keys are no longer functional in Vaultwarden. Bitwarden introduced Universal Prompt support in release 2024.2.3 on March 5th.

Duo's Universal Prompt uses the OIDC Authorization Code flow. This flow requires us to pack information about the authentication we want to be prompted for MFA into an authorization request (a JWT signed with a secret key provided by Duo) and send the user to their service, where MFA is performed. The user is then returned to our service with a code that we use to call Duo's API and obtain the result of the MFA. Once we validate the information passed back by the user's client and returned by Duo's service, we can log the user in.

Detailed documentation is located at [https://duo.com/docs/oauthapi](https://duo.com/docs/oauthapi)

The web vault handles receiving the redirection from Duo. It has a 'connector' page responsible for communicating the authorization code back to the user's true client so it can be provided in a call to Vaultwarden's API.

### Major additions/changes

- Added a crate, duo_oidc, which implements the Duo Universal Prompt MFA flow.
- Added a database table, twofactor_duo_ctx, migrations, and a model to represent it. This table stores information about in-progress Duo MFA attempts for validation while an authenticating user completes MFA on Duo's service.
- Added an internal task to clear the twofactor_duo_ctx table of incomplete Duo MFA attempts.
- Added a configuration option to globally force Vaultwarden to use the legacy Duo traditional prompt flow instead of the universal prompt. 

### Specific Review Items

Beyond what you'd normally check, I'd like to highlight a few decisions I made that you should probably consider in your review.

- The OIDC 'nonce' parameter is optional and intended to harden against replay attacks. The best practice is to bind the nonce to a specific authentication attempt using something like an HttpOnly cookie. Cookies don't seem to be an option without significant changes to Vaultwarden, let alone the Bitwarden clients. Instead of not leveraging it, I tried to win back some of the hardening the parameter is intended to provide. I ended up with a strategy where a random nonce is generated, combined with the `device_identifier` reported by the client, and hashed to produce the OIDC nonce sent to Duo in the authorization request. Then, the `device_identifier` reported by the user's client when providing the authorization code is hashed with the saved nonce, and the resulting hash is compared to the OIDC nonce Duo reports when we call their API for the MFA result.
- Duo's official client libraries for the Universal Prompt pin hard-coded CA certs from Digicert, SecureTrust, and Amazon. Duo's implementation documentation didn't make any recommendations around pinning certs, and it looks like some of them are root CA certs (the pinning of which is controversial), so I did not implement this. 
- Instead of copying or reimplementing the original duo crate's `get_duo_keys_email` function, which fetches the Duo keys from the database/config for a given user, I set its visibility to `pub(crate)` and re-used it. 

### Testing

Tested on the following clients; everything is working well:
- Vaultwarden-patched web vault
- Windows, macOS, and Linux (Flatpak) desktop clients
- Android app (as obtained from Google play) 
- iOS app (as obtained from the Apple app store)
- Chrome and Firefox browser extensions

The only major client issue I found was with the unpackaged AppImage Linux desktop client. When authenticating users on desktop clients, the connector in the web vault opens a `bitwarden://` link, which my system refused to use the AppImage to open. It looks like the AppImage client either doesn't or can't register itself as the `x-scheme-handler` for `bitwarden`, preventing the redirect connector from handing off to the client. It worked fine with the Flatpak packaged Bitwarden client, so I'm chalking it up to my not configuring the AppImage client correctly. 

There is also a known issue upstream in the web vault where the redirect connector shows 'successful authentication', does not automatically close, and the JS 'close' button does not work. See [https://github.com/bitwarden/clients/issues/8554](https://github.com/bitwarden/clients/issues/8554)

I also tested this while running PostgreSQL and MariaDB using the versions packaged in Debian 12; no issues.

BlackDex commented:
Thanks @0x0fbc.

On first glance it looks like you need to rebase already, but besides that it looks ok. No deep checking done.

The only thing that worries me is the collation for the `state`. That is going to give issues i think. Especially when the collation was wrong before and people need to update it. That will change it for that column too.

Isn't there a way to not have this as a key? Or less larger in size. 
(I have not yet looked into this in a functional way, that is why i ask)

0x0fbc commented:
@BlackDex, I reduced the column sizes and removed the collation override I had for MariaDB. I also added the changes suggested by Clippy, did another rustfmt, synced my fork, and rebased.

quexten commented:
@0x0fbc 

> The only major client issue I found was with the unpackaged AppImage Linux desktop client. When authenticating users on desktop clients, the connector in the web vault opens a bitwarden:// link, which my system refused to use the AppImage to open. It looks like the AppImage client either doesn't or can't register itself as the x-scheme-handler for bitwarden, preventing the redirect connector from handing off to the client. It worked fine with the Flatpak packaged Bitwarden client, so I'm chalking it up to my not configuring the AppImage client correctly.

This is an upstream client bug and unrelated to vaultwarden. Both snap and appimage builds have (different) issues with the url handlers at the moment. If I recall correctly from debugging, on AppImage the desktop file does not get created anymore. It needs to be created with the

BlackDex commented on src/api/core/two_factor/duo_oidc.rs:
```diff
@@ -0,0 +1,526 @@
+};
+use url::Url;
+
+// Duo OIDC Auth API URL constants. Defined as macros, so they can be passed into format!()
+#[allow(non_snake_case)]
+macro_rules! HEALTH_ENDPOINT {
+    () => {
+        "https://{}/oauth/v1/health_check"
+    };
+}
+#[allow(non_snake_case)]
+macro_rules! AUTHZ_ENDPOINT {
+    () => {
+        "https://{}/oauth/v1/authorize"
+    };
+}
+#[allow(non_snake_case)]
+macro_rules! API_HOST_FMT {
+    () => {
+        "https://{}"
+    };
+}
+#[allow(non_snake_case)]
+macro_rules! TOKEN_ENDPOINT {
+    () => {
+        "https://{}/oauth/v1/token"
+    };
+}
```
Is there a specific reason to use macro's here?
These _constants_ are only used once.
I think it is better to put these strings directly into the `format!()` instead of using macro's here.

0x0fbc replied:
No objection. I've moved them directly into the `format!()`s where they're used.

They're laid out like this purely as a style choice. I figured if someone came along to this file in the future, it would be helpful to see the endpoints in Duo's API we're interacting with all in one place. I assume you're familiar with the details already, but for my own catharsis I'll leave a technical explanation below as to why they ended up in macros.

---

`format!()` uses the contents of the format string at compile time, so it needs a reference to the format string that exists during macro expansion. Because rustc does macro expansion before name resolution (as macros themselves can generate new names), `format!()` can't take a pointer to const str defined like `const fmt: &str = "Hello, {}!"`. However, we can use a macro that expands to a string literal instead. This is because macros can be nested, and macro expansion does its own limited name resolution step for imports and other macros. Thus, if you want a format string defined somewhere other than the first argument of `format!()`, or if the same format string needs to be used by multiple `format!()`s, it needs to come from a macro.

BlackDex reviewed (changes requested):
Overall looks nice, and it works fine too.
Just some small items i saw from my side.

BlackDex commented:
btw@0x0fbc, sorry for the delay in testing. Also forgive me for pushing some changes, but i forgot the reqwest changes and those didn't popped up as a merge conflict, so i wanted to fix those.

0x0fbc commented:
@BlackDex 

No worries! I figured dealing with the beta client API changes over the last month was much more urgent. Thanks for taking care of the reqwest change. I'll get right on your callouts.

0x0fbc commented on src/api/identity.rs:
```diff
@@ -582,18 +605,36 @@ async fn _json_err_twofactor(providers: &[i32], user_uuid: &str, conn: &mut DbCo
             }
 
-            Some(TwoFactorType::Duo) => {
+            Some(TwoFactorType::Duo | TwoFactorType::OrganizationDuo) => {
```
If I recall correctly, I chose to do this because the clients treat Duo and OrganizationDuo as equivalent. I noticed there wasn't any support for it, and my thoughts at the time were that if a client somehow tried to do OrganizationDuo with us anyway, we could throw them into the Duo logic we could handle. The client wouldn't know the difference.

Coming back to it now, I'd much rather 'fail closed' and not try to let their personal Duo configuration handle the authentication if this occurs.

0x0fbc commented on src/api/identity.rs:
```diff
@@ -518,8 +520,24 @@ async fn twofactor_auth(
         Some(TwoFactorType::YubiKey) => yubikey::validate_yubikey_login(twofactor_code, &selected_data?).await?,
-        Some(TwoFactorType::Duo) => {
-            duo::validate_duo_login(data.username.as_ref().unwrap().trim(), twofactor_code, conn).await?
+        Some(TwoFactorType::Duo | TwoFactorType::OrganizationDuo) => {
```
Removed - [see discussion here](https://github.com/dani-garcia/vaultwarden/pull/4637/files/3d00574621cebb94906f6bc7d3de46402c385458#r1685500019)

0x0fbc commented:
@BlackDex

I completed the requested changes and verified that everything is still working fine.

BlackDex reviewed (approved):
Looks good too me.

Seems to work fine during my tests.

dani-garcia reviewed (commented):
Overall LGTM though I haven't had time to test that it works, just some small comments.

One question, seeing as the Duo traditional prompt is end of life and this is the replacement, is there any special case where someone might want to keep using the old iframe variant? 

Just trying to think for how long we want to support the iframe version in Vaultwarden. If this is a drop in replacement with no downsides, I imagine we can keep support for the traditional prompt for a couple of versions until all the kinks are ironed out and then drop it.

BlackDex commented on src/api/identity.rs:
```diff
@@ -588,12 +611,30 @@ async fn _json_err_twofactor(providers: &[i32], user_uuid: &str, conn: &mut DbCo
+                        let (signature, host) = duo::generate_duo_signature(&email, conn).await?;
+                        result["TwoFactorProviders2"][provider.to_string()] = json!({
+                            "Host": host,
+                            "Signature": signature,
```
I tested it also with the native Android app. Seems to work fine like this.

BlackDex commented:
Retesting the old _iframe_ that still works until somewhere in August if I'm correct. Or maybe a bit longer.

0x0fbc commented on src/db/models/two_factor_duo_context.rs:
```diff
@@ -0,0 +1,86 @@
+
+    pub async fn purge_expired_duo_contexts(conn: &mut DbConn) {
+        for context in Self::find_expired(conn).await {
+            if context.exp < Utc::now().timestamp() {
```
Good catch; that is indeed redundant. I'll remove it.

0x0fbc commented:
Duo hasn't been clear about when they're truly dropping the iframe prompt. For some Duo tenancies (mine included), it's continuing to work; for others, it's not. There's obviously no case to support it after Duo shuts their iframe prompt endpoints down, but I haven't seen any particular date for when they're removing or restricting the endpoints. September 30th, 2024, is the date Duo mentions dropping support for all exceptions that don't have more specific transition dates in the future. I think the Vaultwarden version releasing after that date would be a good time to drop the support. 

The special case I had in mind when choosing to leave in the existing iframe prompt functionality is fairly niche. If I'm not mistaken, the iframe prompt works without Vaultwarden needing to communicate with Duo's API directly, but the new OIDC prompt does not. Someone who's deployed Vaultwarden in a restrictive environment may need to set up firewall exceptions or change other configuration to transition. If these changes would be time-consuming for them, I wanted to avoid forcing them to choose between disabling Duo for their Vaultwarden deployment or downgrading to restore their service while they made whatever arrangements they needed to get the OIDC prompt working.

0x0fbc commented:
Also, I've synced my fork and rebased.

dani-garcia reviewed (approved):
LGTM! Thanks @0x0fbc for the explanations

Gerardv514 commented:
Once merged, are these automatically in the testing tag already?

BlackDex commented:
Once merged it takes mostly less then 1 hour to build all the architectures and create a container.
You can see the status here: https://github.com/dani-garcia/vaultwarden/actions/workflows/release.yml

I just noticed that the alpine builds failed because GitHub killed that container. Those are building right now.

Gerardv514 commented:
Sorry just realized I didn’t post my version.

Server diagnostics has this version: 1.31.0-b4b27019

Is there a place here to go to find out if b4b27019 is latest testing?

GeorgeCastanza commented:
I'm not sure this is **complete**, but Duo is still asking for an update.  I can login without 2FA, which is obviously wrong.

BlackDex commented:
That is not something we can do, you need to configure that via your duo admin panel.

Gerardv514 commented:
Works for me. There is something on duo admin side that you have to do. There’s also an option to use new universal prompt there so make sure that is on.
